server/pulp/server/controllers/repository.py

```diff
@@ -816,6 +816,149 @@ def sync(repo_id, sync_config_override=None, scheduled_call_id=None):
+    """
+    last_sync = conduit.last_sync()
+
+    if last_sync is None:
+        return False
+
+    # convert the iso8601 datetime string to a python datetime object
+    last_sync = dateutils.parse_iso8601_datetime(last_sync)
+    repo_importer = model.Importer.objects.get_or_404(repo_id=repo_id)
+    # the timestamp of the last configuration change
+    last_updated = repo_importer.last_updated
+
+    # check if a configuration change occurred after the most recent sync
+    if last_updated is not None:
+        if last_sync < last_updated:
+            return True
+
+    return False
+
+
+def check_override_config_change(repo_id, call_config):
+    """
+    Checks whether the override config is different from the override config on
+    the previous sync.
+
+    :param repo_id:     identifies the repo to sync
+    :type  repo_id:     str
+    :param call_config: Plugin Call Configuration
+    :type  call_config: pulp.plugins.config.PluginCallConfiguration
+
+    :return:  Whether the override config is different from the override config on
+    the previous sync.
+    :rtype:   boolean
+    """
+    repo_importer = model.Importer.objects.get_or_404(repo_id=repo_id)
+
+    # check if the override config is different from the last override config,
+    # excluding the 'force_full' key.  otherwise using the --force-full flag
+    # would always trigger a full sync on the next sync too
+    prev_config = repo_importer.last_override_config.copy()
+    current_config = call_config.override_config.copy()
+
+    prev_config.pop('force_full', False)
+    current_config.pop('force_full', False)
+
+    return prev_config != current_config
+
+
+def check_perform_full_sync(repo_id, conduit, call_config):
+    """
+    Performs generic checks to determine if the sync should be a "full" sync.
+    Checks if the "force full" flag has been set, if content has been removed
+    since the last sync, and whether the configuration has changed since the
+    last sync (including override configs).
+
+    Plugins may want to perform additional checks beyond the ones appearing here.
+
+    :param repo_obj: repository object
+    :type  repo_obj: pulp.server.db.model.Repository
+    :param conduit: allows the plugin to interact with core pulp
+    :type  conduit: pulp.plugins.conduits.repo_publish.RepoPublishConduit
+    :param call_config: Plugin Call Configuration
+    :type  call_config: pulp.plugins.config.PluginCallConfiguration
+
+    :return:  Whether a full sync needs to be performed
+    :rtype:  boolean
+    """
+    force_full = check_force_full(call_config)
+    not_synced = check_no_previous_sync(conduit)
+    content_removed = check_unit_removed_since_last_sync(conduit, repo_id)
+    config_changed = check_config_updated_since_last_sync(conduit, repo_id)
+    override_config_changed = check_override_config_change(repo_id, call_config)
+
+    return force_full or not_synced or content_removed or config_changed or override_config_changed
```

Comment:
i see your point, but i don't think that difficulty of testing the function should influence/dictate of how the actual function should be written/look like.
I would like hear also someone else opinion about this, because due to the fact you've slitted the checks into multiple functions they make repeated calls to the db.